### PR TITLE
Improve accuracy of diagnostic lines in Elixir compile task

### PR DESCRIPTION
**Pull request by JakeBecker**
Fixes two scenarios in which the diagnostic line numbers are not correct:

  * `UndefinedFunctionError`s would cause the line number to be `nil`, but we can determine the line number from the stack trace
  * When the error originates in a file other than the one that was being compiled, the line number will be the line at which the other file had an exception, which is confusing. Instead, it now sets the line number to `nil`.

### CODE REVIEW

**Comment by josevalim** [comment]
Thanks @JakeBecker! I am not sure if we should generalize the logic like that because there is no evidence this is a general property or something specific to the function clause error. Therefore I would add a clause that is specific to it. Maybe something like this:

    defp get_line(:undef, [{_, _, _, []}, {_, _, _, info} | _]), do: Keyword.get(info, :line)

Thoughts?

PS: I will review the other PR hopefully today.

**Comment by JakeBecker** [comment]
That works, though you still have to check if the file matches in case it's being raised from a file other than the one being compiled. I made the change.

File: lib/elixir/lib/kernel/parallel_compiler.ex [hunk 1]
```diff
@@ -410,13 +410,29 @@ defmodule Kernel.ParallelCompiler do
...
+  end
+
+  defp get_line(file, :undef, [{_, _, _, []}, {_, _, _, info} | _]) do
+    if Keyword.get(info, :file) == to_charlist(file) do
+      Keyword.get(info, :line)
+    end
+  end
+
+  defp get_line(file, :undef, [{_, _, _, info} | _]) do
+    if Keyword.get(info, :file) == to_charlist(file) do
```
Comment (josevalim) [inline lib/elixir/lib/kernel/parallel_compiler.ex]: Do we still need those files check (this and the one above)?

Comment (josevalim) [reply lib/elixir/lib/kernel/parallel_compiler.ex]: Ok, you have already replied why we need it. Thanks!

**Comment by JakeBecker** [comment]
Fixed a typo, the second pattern should match any reason, not just `:undef`.

File: lib/mix/test/mix/tasks/compile.elixir_test.exs [hunk 2]
```diff
@@ -438,4 +438,50 @@ defmodule Mix.Tasks.Compile.ElixirTest do
       end)
     end
   end
+
+  test "does not use incorrect line number when error originates in another file" do
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/compile.elixir_test.exs]: Can you please move this to the parallel_compiler_test.exs? I don't think we should test it through mix. You can use `Path.join(tmp_path(), "parallel_compiler_undef.exs")` in there to generate a path in the temp directory where you will write the file.

File: lib/mix/test/mix/tasks/compile.elixir_test.exs [hunk 3]
```diff
@@ -438,4 +438,50 @@ defmodule Mix.Tasks.Compile.ElixirTest do
...
+      defmodule B do
+        def fun(arg) do
+          A.fun(arg)
+          :ok
+        end
+      end
+      B.fun(:not_a_number)
+      """)
+
+      import ExUnit.CaptureIO
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/compile.elixir_test.exs]: When you move this test to paralle_compiler_test.exs, you will no longer need this import.

File: lib/mix/test/mix/tasks/compile.elixir_test.exs [hunk 4]
```diff
@@ -438,4 +438,50 @@ defmodule Mix.Tasks.Compile.ElixirTest do
...
+      UndefinedFunctionError
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/compile.elixir_test.exs]: This seems to be a leftover? :)

File: lib/mix/test/mix/tasks/compile.elixir_test.exs [hunk 5]
```diff
@@ -438,4 +438,50 @@ defmodule Mix.Tasks.Compile.ElixirTest do
...
+      capture_io(fn ->
+        file = Path.absname("lib/b.ex")
+        assert {:error, [%{file: ^file, position: nil}]} = Mix.Tasks.Compile.Elixir.run([])
+      end)
+    end
+  end
+
+  test "gets correct line number for UndefinedFunctionError" do
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/compile.elixir_test.exs]: Please move this test and remove the import below as well!

**Comment by josevalim** [comment]
Thanks @JakeBecker. The code is perfect, now we just need to move the tests around and we are good to go. Btw, in case you don't know, you can run the tests for any invidual file like this:

    elixir -r lib/elixir/test/elixir/test_helper.exs lib/elixir/test/elixir/kernel/parallel_compiler_test.exs

File: lib/mix/test/mix/tasks/compile.elixir_test.exs (hunk 4 above)
Comment (JakeBecker) [reply lib/mix/test/mix/tasks/compile.elixir_test.exs]: Whoops!

**Comment by JakeBecker** [comment]
Thanks for the feedback @josevalim . I made the changes, and I discovered that it wasn't able to get the error line from the stack trace if absolute paths are passed into ParallelCompiler. I made a small change to fix that too.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**